dwave/cloud/solver.py

```diff
@@ -739,6 +749,149 @@ def upload_cqm(self, cqm):
         return self.upload_problem(cqm)
 
 
+class NLSolver(BaseUnstructuredSolver):
+    """NL solver interface.
+
+    This class provides an :term:`NL model` sampling method and encapsulates
+    the solver description returned from the D-Wave cloud API.
+
+    Args:
+        client (:class:`~dwave.cloud.client.Client`):
+            Client that manages access to this solver.
+
+        data (`dict`):
+            Data from the server describing this solver.
+    """
+
+    _handled_problem_types = {"nl"}
+    _handled_encoding_formats = {"binary-ref"}
+
+    def _encode_problem_for_upload(self,
+                                   model: typing.Union['dwave.optimization.Model', io.IOBase],
+                                   **kwargs
+                                   ) -> io.IOBase:
+        encode_params = {k: v for k, v in kwargs.items()
+                         if k in {'max_num_states', 'only_decision'}}
+        try:
+            data = model.to_file(**encode_params)
+        except Exception as e:
+            logger.debug("NL model serialization failed with %r, "
+                         "assuming data already encoded.", e)
+            # assume `model` given as file, ready for upload
+            data = model
+        else:
+            logger.debug("Problem (model) encoded with encode_params=%r for upload in %r",
+                        encode_params, data)
+
+        return data
+
+    def sample_bqm(self, bqm, label=None, **params):
+        """Use just for testing."""
+        try:
+            import dimod
+        except ImportError: # pragma: no cover
+            raise RuntimeError("Can't sample from 'bqm' without dimod. "
+                               "Re-install the library with 'bqm' support.")
+        try:
+            from dwave.optimization.generators import _from_constrained_quadratic_model
+        except ImportError: # pragma: no cover
+            raise RuntimeError("Can't sample from nonlinear model without dwave-optimization. "
+                               "Re-install the library with 'nlm' support.")
+
+        # cqm to model generator requires binary bqm
+        bqm.change_vartype(dimod.BINARY, inplace=True)
+
+        # TODO: simplify when a public BQM -> Model generator is added to dwave-optimization
+        cqm = dimod.CQM.from_bqm(bqm)
+        nlm = _from_constrained_quadratic_model(cqm)
+
+        return self.sample_nlm(nlm, label=label, **params)
```

Comment:
Should this return a sample set? I guess if it's just for testing we could change it later.

Reply:
No, all `Solver.sample_*` methods return `dwave.cloud.computation.Future`. I missed annotating the rtype here. Will fix.